proto/prost/private/protoc_wrapper.rs

```diff
@@ -0,0 +1,1047 @@
+fn find_generated_rust_files(out_dir: &Path) -> BTreeSet<PathBuf> {
+    let mut all_rs_files: BTreeSet<PathBuf> = BTreeSet::new();
+    for entry in fs::read_dir(out_dir).expect("Failed to read directory") {
+        let entry = entry.expect("Failed to read entry");
+        let path = entry.path();
+        if path.is_dir() {
+            for f in find_generated_rust_files(&path) {
+                all_rs_files.insert(f);
+            }
+        } else if let Some(ext) = path.extension() {
+            if ext == "rs" {
+                all_rs_files.insert(path);
+            }
+        } else if let Some(name) = path.file_name() {
+            if name == "_" {
+                let rs_name = path.parent().expect("Failed to get parent").join("_.rs");
+                fs::rename(&path, &rs_name).unwrap_or_else(|err| {
+                    panic!("Failed to rename file: {err:?}: {path:?} -> {rs_name:?}")
+                });
+                all_rs_files.insert(rs_name);
+            }
+        }
+    }
+
+    all_rs_files
+}
+
+/// Rust module definition.
+#[derive(Debug, Default)]
+struct Module {
+    /// The name of the module.
+    name: String,
+
+    /// The contents of the module.
+    contents: String,
+
+    /// The names of any other modules which are submodules of this module.
+    submodules: BTreeSet<String>,
+}
+
+/// Generate a lib.rs file with all prost/tonic outputs embeeded in modules which
+/// mirror the proto packages. For the example proto file we would expect to see
+/// the Rust output that follows it.
+///
+/// ```proto
+/// syntax = "proto3";
+/// package examples.prost.helloworld;
+///
+/// message HelloRequest {
+///     // Request message contains the name to be greeted
+///     string name = 1;
+/// }
+//
+/// message HelloReply {
+///     // Reply contains the greeting message
+///     string message = 1;
+/// }
+/// ```
+///
+/// This is expected to render out to something like the following. Note that
+/// formatting is not applied so indentation may be missing in the actual output.
+///
+/// ```ignore
+/// pub mod examples {
+///     pub mod prost {
+///         pub mod helloworld {
+///             // @generated
+///             #[allow(clippy::derive_partial_eq_without_eq)]
+///             #[derive(Clone, PartialEq, ::prost::Message)]
+///             pub struct HelloRequest {
+///                 /// Request message contains the name to be greeted
+///                 #[prost(string, tag = "1")]
+///                 pub name: ::prost::alloc::string::String,
+///             }
+///             #[allow(clippy::derive_partial_eq_without_eq)]
+///             #[derive(Clone, PartialEq, ::prost::Message)]
+///             pub struct HelloReply {
+///                 /// Reply contains the greeting message
+///                 #[prost(string, tag = "1")]
+///                 pub message: ::prost::alloc::string::String,
+///             }
+///             // @protoc_insertion_point(module)
```

Comment:
How do you feel about allowing adding code here? (Obviously can be a follow-up if you think it makes sense to allow)

Reply:
That sounds good to me. I haven't thought much about what would need to change to make that work. I'd definitely want to defer that work to another pr / issue. Should I create an issue for that?

Reply:
Feel free to create an issue, or just ignore for now and one of us will get to it at some point :)

I'm imagining this taking the form roughly of a new attribute on `rust_*_library` along the lines of:
```
extra_per_package_code = {
  "my.proto.package": "path/to:extra.rs",
}
```

or something like that? But very open to ideas!

Reply:
I created a feature request for this here: https://github.com/bazelbuild/rules_rust/issues/2045